### PR TITLE
post(tips): Tip 3, [hidden] vs Bootstrap's .help-block

### PR DESCRIPTION
**Pull request by clacote**
Not sure this is very useful. But I did my part of the job...

### CODE REVIEW

File: _posts/2016-03-22-ninja-tips-3-hidden-vs-bootstrap-help-block-english.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    <span class="help-block"
+          *ngIf="loginForm.find('password').dirty
+                 && loginForm.find('password').hasError('required')">
+      Password is required
+    </span>
+
+Et voilà&nbsp;!
+Of course, we needed to inverse the initial condition of `[hidden]="..."` (but that's Boole Algebra 101),
+and we replaced the `pristine` test by a `dirty` test, which are the exact inverse.
```
Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-hidden-vs-bootstrap-help-block-english.md]: which are the exact inverse -> which is the exact opposite?

"inverse the initial condition" sounds weird... To negate maybe?

File: _posts/2016-03-22-ninja-tips-3-hidden-vs-bootstrap-help-block-english.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
+---
+layout: post
+title: Ninja Tips 3 - [hidden] vs Bootstrap's help-block
+author: clacote
+tags: ["Angular 2", "bootstrap", "html", "css"]
+description: "Tip about form validation with Angular 2 and Bootstrap"
+---
+
+Like everybody, you code enterprise software for a living
+(because everybody know you can't earn your life by building videogames).
```
Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-hidden-vs-bootstrap-help-block-english.md]: "earn a living" or "make a living" is better I think.

File: _posts/2016-03-22-ninja-tips-3-hidden-vs-bootstrap-help-block-english.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+And as you're not very good at front design, you enjoy using CSS frameworks
```
Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-hidden-vs-bootstrap-help-block-english.md]: at front design -> at designing web pages

File: _posts/2016-03-22-ninja-tips-3-hidden-vs-bootstrap-help-block-english.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      </span>
+    </div>
+
+    <button type="submit" class="btn btn-primary" [disabled]="!loginForm.valid">
+      Sign me in
+    </button>
+
+What have we done here?
+
+* We apply thanks to `ngClass` directive the Bootstrap's `.has-error` class on `div.form-group` if the `password` field:
```
Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-hidden-vs-bootstrap-help-block-english.md]: We apply, thanks to ... directive,

**Comment by cexbrayat** [comment]
:+1:

**Comment by clacote** [comment]
@cexbrayat review fixed.

@jnizet: will you have the time to review this before tomorrow morning (its hypothetical publication date)?

**Comment by jnizet** [comment]
I'll try reading it at noon.

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      <label for="password">Password</label>
+      <input id="password" type="password"
+             class="form-control" ngControl="password" required>
+      <span class="help-block">
+        Password is required
+      </span>
+    </div>
+
+But as soon as this form is displayed, this help text is displayed,
+even if the user has not type anything yet.
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: typed

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+At Ninja Squad, we like to have those hints and error messages be displayed only
+when the user has started to input something, to keep a clean form when entering the page.
+You may then leverage the validation capabilities of a super-charged Angular&nbsp;2 form:
+
+    <div class="form-group"
+         [ngClass]="{
+            'has-error': loginForm.find('password').dirty
+                         && !loginForm.find('password').valid }">
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: I dislike this `loginForm.find('password')` in the view. And even more when it's duplicated. Why not assume that the password control is a field of the component, and just do `password.dirty && !password.valid`?

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+      </span>
+    </div>
+
+    <button type="submit" class="btn btn-primary" [disabled]="!loginForm.valid">
+      Sign me in
+    </button>
+
+What have we done here?
+
+* We apply, thanks to `ngClass` directive, the Bootstrap's `.has-error` class on `div.form-group` if the `password` field:
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: thanks to **the** ngClass directive. It's strange to introduce the section with "what have we done", and then continue in the present tense.

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  1. is `dirty`, i.e. modified by the user;
+  2. has a validation error.
+* We enable the submit button only if the form is globally valid, thanks to HTML `disabled` attribute.
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: thanks to **the** HTML `disabled` attribute. But, IIRC, it's not really an HTML attribute, but rather a DOM property. @cexbrayat do you confirm?

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+* We hide the `span.help-block` element thanks to the HTML&nbsp;5 `hidden` attribute, if:
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: Same here, it's a DOM property

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+    <div class="form-group"
+         [ngClass]="{
+            'has-error': loginForm.find('password').dirty
+                         && !loginForm.find('password').valid }">
+      <label for="password">Password</label>
+      <input id="password" type="password"
+             class="form-control" ngControl="password">
+      <span class="help-block"
+            [hidden]="loginForm.find('password').pristine
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: Same here

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+          *ngIf="loginForm.find('password').dirty
+                 && loginForm.find('password').hasError('required')">
+      Password is required
+    </span>
+
+Et voilà&nbsp;!
+Of course, we needed to negate the initial condition of `[hidden]="..."` (but that's Boole Algebra 101),
+and we replaced the `pristine` test by a `dirty` test, which is the exact opposite.
+
+Thanks to have taken the time to listen to me until then (I feel better now).
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: That dosn't look like correct English to me. Maybe something like

> Thanks for the time you spent listening to me. I feel better now.

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-francais.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    <button type="submit" class="btn btn-primary" [disabled]="!loginForm.valid">
+      Me connecter
+    </button>
+
+Qu'avons-nous fait ici&nbsp;?
+
+* On applique grâce à la directive `ngClass` la classe Bootstrap `.has-error` sur la `div.form-group` si le champ `password`&nbsp;:
+  1. est `dirty`, c'est à dire modifié par l'utilisateur&nbsp;;
+  2. a une erreur de validation.
+* On active le bouton de soumission que si le formulaire est valide, grâce à l'attribut HTML `disabled`.
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-francais.md]: On n'active

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-francais.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+* On masque le `span.help-block` grace à l'attribut HTML&nbsp;5 `hidden`, si&nbsp;:
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-francais.md]: grâce

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-francais.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    <span class="help-block"
+          *ngIf="loginForm.find('password').dirty
+                 && loginForm.find('password').hasError('required')">
+      Password is required
+    </span>
+
+Et voilà&nbsp;!
+Il nous a fallu inverser la condition initialement placée dans `[hidden]="..."` (mais c'est de l'algèbre de Boole niveau CP), et on a remplacé le test de `pristine` par un test de `dirty`, qui sont les opposés.
+
+Merci d'avoir suivi ce long cheminement pour me permettre de vous expliquer cette anecdote, je me sens mieux d'avoir pu vider mon cœur. Vous pouvez reprendre de votre côté une activité normale (comme la lecture de [notre ebook sur Angular&nbsp;2](https://books.ninja-squad.com/angular2) par exemple).
```
Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-francais.md]: vider mon sac? sortir ce que j'avais sur le coeur?

J'enlèverais "de vote côté", pour que la référence aux guignols soit plus flagrante.

**Comment by jnizet** [comment]
J' n'ai pas remis dans le post en français toutes les remarques concernant le code, et la terminologie attribut vs. propriété DOM, mais elles s'appliquent évidemment.

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md (hunk 8 above)
Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: Yep, it's a DOM property

File: _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md (hunk 9 above)
Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: Yes, same thing. I think hidden is fairly old, but I did not check

Comment (clacote) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: Not so old: it appears with HTML 5.

Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: Ho really? Maybe the hidden attribute, but the hidden DOM property is older right?

Comment (clacote) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: I have no idea... And I don't really know where to look. What is the reference?

Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: I think you're right actually: there is no hidden mentioned in the DOM [level 2 spec](https://www.w3.org/TR/DOM-Level-2-HTML/html.html#ID-58190037) but it mentioned in the [HTML5 spec](https://www.w3.org/TR/html5/dom.html#htmlelement). And I don't think there are other specs for the HTML part of the DOM between those two. I've learnt something.

Comment (jnizet) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: Does that mean that the syntax `[hidden]="..."`can set HTML attributes in addition to DOM properties?

Comment (cexbrayat) [inline _posts/2016-03-22-ninja-tips-3-angular-validation-hidden-vs-bootstrap-help-block-english.md]: De ce que j'en sais, tout attribut est le reflet (parfois partiel) d'une propriété équivalente dans le DOM (à qq [exceptions de nom près](https://github.com/angular/angular/blob/e7470d557df0400a812d84551524970cb5f12eee/modules/angular2/src/platform/server/parse5_adapter.ts#L20)). 

`[hidden]` interagit avec la propriété DOM `hidden`, mais ne set pas l'attribut `hidden` à ma connaissance. Par contre, est-ce que le navigateur le fait automatiquement? En testant en Vanilla JS, on dirait que oui. 

capture d ecran 2016-03-21 a 18 32 27

**Comment by clacote** [comment]
@jnizet review fixed in 3669c57.
I've written "`hidden` HTML 5 global attribute", "`disabled` DOM property".

**Comment by jnizet** [comment]
:+1:

**Comment by acrepet** [comment]
Nice

**Pull request merged**